test_runner/regress/test_replication_start_subxid_overflow.py

```diff
@@ -0,0 +1,235 @@
+from fixtures.log_helper import log
+from fixtures.neon_fixtures import NeonEnv, wait_for_last_flush_lsn, wait_replica_caughtup
+
+
+def test_replication_start_subxid_overflow(neon_simple_env: NeonEnv):
+    env = neon_simple_env
+
+    with env.endpoints.create_start(branch_name="main", endpoint_id="primary") as primary:
+        with primary.connect() as p_con:
+            with p_con.cursor() as p_cur:
+                p_cur.execute("begin")
+                p_cur.execute("create table t(pk serial primary key, payload integer)")
+                p_cur.execute(
+                    """create or replace function create_subxacts(n integer) returns void as $$
+                    declare
+                        i integer;
+                    begin
+                       for i in 1..n loop
+                          begin
+                              insert into t (payload) values (0);
+                          exception
+                              when others then
+                                  raise exception 'caught something';
+                          end;
+                       end loop;
+                   end; $$ language plpgsql"""
+                )
+                p_cur.execute("select create_subxacts(100000)")
+                xid = p_cur.fetchall()[0][0]
+                log.info(f"Master transaction {xid}")
+                with env.endpoints.new_replica_start(
+                    origin=primary, endpoint_id="secondary"
+                ) as secondary:
+                    wait_replica_caughtup(primary, secondary)
+                    with secondary.connect() as s_con:
+                        with s_con.cursor() as s_cur:
+                            # Enforce setting hint bits for pg_class tuples.
+                            # If master's transaction is not marked as in-progress in MVCC snapshot,
+                            # then XMIN_INVALID hint bit will be set for table's 't' tuple makeing it invisible.
```

Comment:
I don't understand this comment about hint bits and pg_class tuples. What happens if you remove this?

Reply:
I have explained it multiple times. Ok, will try to repeat it once again. So what is their scenario of data corruption?
1. Replica receives some uncommitted changes.
2. Transaction is not marked as in progress
3. Vacuum or just arbitrary query traverse through this data and set XMIN_INVALID hint bit (because transaction is not marked as been committed or in-progress)
4. Starting from this moment tuple becomes permanently unvisible even after transition is committed at primary and commit applied by replica.

So this is is why I have added query which iterate through pg_class and set hint bits.

Reply:
That sounds incredibly bad. How can we fix that?

Reply:
Sorry, what is incredibly bad? 
Test? Bug? The way we are fixing it?

This problem is known and actively discussed during several month. And I have noticed at the very beginning that it can cause such kind of error at  replica (replica actually becomes unusable until you restart it).

Original attempt to fix this issue is to make replica get information about running xacts from primary. 
But it cause slowdown of replica startup and this PR was reverted.

This one is yet another attempt to solve this problem. This approach with extracting running xacts from CLOG was actually proposed by you. It seems to be quite simple and reliable ... except this problem Wirth large number of subxacts. I think that this scenario with large number of subjects is quite rare. And the original problem with lack of information about running xacts is quite critical. This is why I suggested to commit this PR even through there are known issues with large number of subxacts.

Right now I now sure about further steps with this PR. SHow we 
1. Find some way to prevent knowns xid overflow error? How? Just decrease threashold from MAX/2 till some smalller?
2. Move this code to extension together with some other Neon changes in xlog.c code?
3. Investigate and find some other way how to deal with sunning xacts instead of extracting them from CLOG?
From my point of view there are only two alternatives:
- make replica to wait subxacts from primary and pink primary to generate running xacts wal record.
- implement your proposal with using CSN at replica.

Reply:
I think we need to have a fix that eliminates the cases of corruption, and incorrect query results. Killing queries with errors is OK, and restarting the replica if you hit some corner is kind of OK too. Silent corruption or incorrect query results is not.

Reply:
> Find some way to prevent knowns xid overflow error? How? Just decrease threashold from MAX/2 till some smalller?

Crucially, if all the transactions don't fit in the known-assigned XIDs array, we should not allow hot standby queries, and fall back waiting until you get the running-xacts record.

And if the the known-assigned XIDs array fills up later during replay, kill the replica with an ERROR. Or could we kill just the running queries, and not allow new queries until we see the running-xacts record?

If we want to be sure that the known-assigned xids array cannot fill up later, how small do we need to make the MAX/2 threshold?